homeassistant/components/zwave_js/__init__.py

```diff
@@ -886,67 +916,52 @@ async def client_listen(
+            raise
         LOGGER.error("Failed to listen: %s", err)
     except Exception as err:  # pylint: disable=broad-except
         # We need to guard against unknown exceptions to not crash this task.
         LOGGER.exception("Unexpected exception: %s", err)
+        if entry.state != ConfigEntryState.LOADED:
+            raise
 
     # The entry needs to be reloaded since a new driver state
     # will be acquired on reconnect.
     # All model instances will be replaced when the new state is acquired.
-    if should_reload:
-        LOGGER.info("Disconnected from server. Reloading integration")
+    if not hass.is_stopping:
+        LOGGER.debug("Disconnected from server. Reloading integration")
         hass.async_create_task(hass.config_entries.async_reload(entry.entry_id))
 
 
-async def disconnect_client(hass: HomeAssistant, entry: ConfigEntry) -> None:
-    """Disconnect client."""
-    data = hass.data[DOMAIN][entry.entry_id]
-    client: ZwaveClient = data[DATA_CLIENT]
-    listen_task: asyncio.Task = data[DATA_CLIENT_LISTEN_TASK]
-    start_client_task: asyncio.Task = data[DATA_START_CLIENT_TASK]
-    driver_events: DriverEvents = data[DATA_DRIVER_EVENTS]
-    listen_task.cancel()
-    start_client_task.cancel()
-    platform_setup_tasks = driver_events.platform_setup_tasks.values()
-    for task in platform_setup_tasks:
-        task.cancel()
-
-    tasks = (listen_task, start_client_task, *platform_setup_tasks)
-    await asyncio.gather(*tasks, return_exceptions=True)
-    for task in tasks:
-        with suppress(asyncio.CancelledError):
-            await task
-
-    if client.connected:
-        await client.disconnect()
-        LOGGER.info("Disconnected from Zwave JS Server")
-
-
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a config entry."""
     info = hass.data[DOMAIN][entry.entry_id]
     client: ZwaveClient = info[DATA_CLIENT]
     driver_events: DriverEvents = info[DATA_DRIVER_EVENTS]
 
-    tasks: list[Coroutine] = [
-        hass.config_entries.async_forward_entry_unload(entry, platform)
-        for platform, task in driver_events.platform_setup_tasks.items()
-        if not task.cancel()
-    ]
+    setup_tasks: list[Awaitable] = []
+    unload_tasks: list[Awaitable] = []
 
-    unload_ok = all(await asyncio.gather(*tasks)) if tasks else True
+    for platform, task in driver_events.platform_setup_tasks.items():
+        if not task.cancel():
+            unload_tasks.append(
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+            )
+        else:
+            setup_tasks.append(task)
 
-    if client.connected and client.driver:
-        await async_disable_server_logging_if_needed(hass, entry, client.driver)
-    if DATA_CLIENT_LISTEN_TASK in info:
-        await disconnect_client(hass, entry)
+    unload_ok = all(await asyncio.gather(*unload_tasks)) if unload_tasks else True
+    await asyncio.gather(*setup_tasks)
```

Comment:
We may need to handle exceptions here so that we don't crash the config entry unload. Although this is now just the config entry unloading and forwarding tasks so it shouldn't really blow up, I think.

Reply:
Being more defensive here is good, given that the integration triggers the unload process when it detects a config change.  In other words, it isn't always part of a user initiated action like a shutdown or reload.  The unload crashed for me in the middle of the night after a power flicker (and there was not config change - but that's a separate transient issue) - my system is on a UPS but most of the zwave devices are not.

So if having an exception leave that function crashes the integration, I'd catch Exception and log it and it that message ask the user to submit it.

Reply:
We're trying to get to solve the exceptions while unloading with this PR. They are not from platform forwarding or unload.